components/brave_wallet/browser/zcash/zcash_shield_sync_service.cc

```diff
@@ -0,0 +1,384 @@
+    ScheduleWorkOnTask();
+    return;
+  } else {
+    // Reorg database so records related to removed blocks are wiped out
+    background_orchard_storage_
+        .AsyncCall(&ZCashOrchardStorage::HandleChainReorg)
+        .WithArgs(account_id_.Clone(), (*tree_state)->height,
+                  (*tree_state)->hash)
+        .Then(base::BindOnce(
+            &ZCashShieldSyncService::OnDatabaseUpdatedForChainReorg,
+            weak_ptr_factory_.GetWeakPtr(), (*tree_state)->height));
+  }
+}
+
+void ZCashShieldSyncService::OnDatabaseUpdatedForChainReorg(
+    uint32_t new_block_height,
+    std::optional<ZCashOrchardStorage::Error> error) {
+  if (error) {
+    error_ = Error{kFailedToUpdateDatabase, error->message};
+    ScheduleWorkOnTask();
+  } else {
+    latest_scanned_block_ = new_block_height;
+    // Reload list of discovered spendable notes and update account info
+    UpdateSpendableNotes();
+  }
+}
+
+void ZCashShieldSyncService::UpdateSpendableNotes() {
+  spendable_notes_ = std::nullopt;
+  background_orchard_storage_.AsyncCall(&ZCashOrchardStorage::GetSpendableNotes)
+      .WithArgs(account_id_.Clone())
+      .Then(base::BindOnce(&ZCashShieldSyncService::OnGetSpendableNotes,
+                           weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashShieldSyncService::OnGetSpendableNotes(
+    base::expected<std::vector<OrchardNote>, ZCashOrchardStorage::Error>
+        result) {
+  if (!result.has_value()) {
+    error_ = Error{kFailedToRetrieveSpendableNotes, result.error().message};
+  } else {
+    spendable_notes_ = result.value();
+    current_sync_status_ = mojom::ZCashShieldSyncStatus::New(
+        latest_scanned_block_.value(), chain_tip_block_.value(),
+        spendable_notes_->size(), GetSpendableBalance());
+  }
+
+  ScheduleWorkOnTask();
+}
+
+void ZCashShieldSyncService::UpdateChainTip() {
+  zcash_rpc_->GetLatestBlock(
+      chain_id_, base::BindOnce(&ZCashShieldSyncService::OnGetLatestBlock,
+                                weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashShieldSyncService::OnGetLatestBlock(
+    base::expected<zcash::mojom::BlockIDPtr, std::string> result) {
+  if (!result.has_value()) {
+    error_ = Error{kFailedToUpdateChainTip, result.error()};
+  } else {
+    chain_tip_block_ = (*result)->height;
+  }
+
+  ScheduleWorkOnTask();
+}
+
+void ZCashShieldSyncService::DownloadBlocks() {
+  zcash_rpc_->GetCompactBlocks(
+      chain_id_, *latest_scanned_block_ + 1,
+      std::min(*latest_scanned_block_ + 1,
+               *latest_scanned_block_ + kScanBatchSize),
+      base::BindOnce(&ZCashShieldSyncService::OnBlocksDownloaded,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void ZCashShieldSyncService::OnBlocksDownloaded(
+    base::expected<std::vector<zcash::mojom::CompactBlockPtr>, std::string>
+        result) {
+  if (!result.has_value()) {
+    error_ = Error{ErrorCode::kFailedToDownloadBlocks, result.error()};
+    ScheduleWorkOnTask();
```

Comment:
Do we need to care about ordering of these tasks being scheduled here or is order guaranteed such that as long as we're not executing out of order we won't find ourselves in weird state changes?

Reply:
Only one step should be executed at the time

Reply:
Ahh so we won't build a queue of scheduled tasks? Sorry this is my first time seeing the scheduler so trying to think through potential dangers we might encounter while learning it. Seems like what you described is non-blocking if I'm understanding your answer correctly.

Reply:
Yes, there is no queue, we use post task to prevent recursive calls on the thread